src/yki/boundary/onr.clj

```diff
@@ -1,23 +1,149 @@
+      (assoc
+        basic-fields
+        :hetu (str/upper-case ssn)
+        :eiSuomalaistaHetua false)
+      (assoc
+        basic-fields
+        :syntymaaika birthdate
+        :identifications (registration+attempt->identifications registration attempt)
+        :eiSuomalaistaHetua true))))
+
+(defn- returned-person-details-match? [registration person-response]
+  (cond
+    ; SSN supplied for person -> trust the OID returned, but log a warning if other details do not match!
+    (has-ssn? registration)
+    (do
+      (when-not (= (:last_name registration)
+                   (person-response "sukunimi"))
+        (log/warn
+          (str "Found OID for registration with hetu, but supplied and returned last names differ. Registration id: "
+               (:registration_id registration)
+               ", OID: "
+               (person-response "oidHenkilo")
+               ", supplied: "
+               (:last_name registration)
+               ", returned: "
+               (person-response "sukunimi"))))
+      true)
+    ; Use birthdate as the next best criteria to gauge if the person in ONR matches the one on our registration form
+    (= (:birthdate registration) (person-response "syntymaaika"))
+    true
+    ; Unmatching birthdates => infer a mismatch and log a warning
+    :else
+    (do
+      (log/warn
+        (str "Found OID but supplied and returned birthdates differ. Registration id: "
+             (:registration_id registration)
+             ", OID: "
+             (person-response "oidHenkilo")
+             ", supplied: "
+             (:birthdate registration)
+             ", returned: "
+             (person-response "syntymaaika")))
+      false)))
+
+(defn response->body [response]
+  (json/read-value (:body response)))
+
+(defn response->status [response]
+  (:status response))
+
 (defprotocol Onr
   (get-or-create-person [this person])
   (get-person-by-ssn [this ssn])
   (get-person-by-oid [this oid]))
 
+(defn- get-or-create-person-with-retries [cas-client onr-url registration max-attempts]
+  (loop [attempt 1
+         oid     nil]
+    (if (< max-attempts attempt)
+      (do
+        (log/error
+          (str "Retries exhausted. Returning OID of last match. Registration id: "
+               (:registration_id registration)
+               ", OID: "
+               oid))
+        oid)
+      (let [onr-person (registration->onr-person registration attempt)
+            response   (cas/cas-authenticated-post cas-client onr-url onr-person)
+            status     (response->status response)]
+        (cond
+
+          ; New person created within ONR -> return the OID returned in response
+          (= 201 status)
+          ((response->body response) "oidHenkilo")
+
+          ; Found an existing person in ONR -> check details and return OID if they look like a reasonable match
+          ; Otherwise try again with slightly altered identification details to avoid using someone else's OID.
```

Comment:
En saanut tästä aivan kiinni, että missä tuo uudelleen yritys tapahtuu ja mitä tarkoittaa tässä tapauksessa "slightly altered identification details"?

Reply:
Jees! Eli tuossa rivillä 133 vertaillaan täsmääkö ONR:stä löytyneen henkilön tiedot niihin mitä ilmoittautumislomakkeella on annettu; jos täsmää niin tästä palautetaan OID ja jos ei niin `(recur (inc attempt) oid)` vie suorituksen takaisin tämän `loop`-formin alkuun. Se mitä tunnistetietoja milläkin kerralla yritetään riippuu tuosta `attempt`in arvosta, toteutus löytyy `registration->onr-person`-funkkarista alkaen riviltä 29.

Reply:
Aa noniin, nyt sain luettua tuota syntaksiakin paremmin, tää selvä